 #! /usr/bin/env python
# -*- coding: utf-8 -*-

# Setup
#

configfile: "config/config.yaml"

# wildcard constraints
#
wildcard_constraints:
    sra_run = "SRR[0-9]{6,9}",
    dataset = "PRJNA\d+"

# include rules
#
include: "rules/common.smk"
include: "rules/get_fastq_files.smk"
include: "rules/align_multi_starsolo.smk"
include: "rules/canonicalQC.smk"
include: "rules/stellarscope.smk"
include: "rules/analysis.smk"
#include: "rules/sample_complete.smk"

rule rna_gex:
    input:
        expand('results/rna_gex_PRJNA587486/align_multi_starsolo/{sample}/Aligned.sortedByCoord.out.bam',
            sample=meta_table[meta_table['dataset'] == 'rna_gex_PRJNA587486'].index
        ),
        expand('results/rna_gex_PRJNA587486/stellarscope/{sample}/poolmodes.txt',
            sample=meta_table[meta_table['dataset'] == 'rna_gex_PRJNA587486'].index
        ),        
    output:
        touch('results/rna_gex/completed.txt')


